controllers/stargate_controller.go

```diff
@@ -0,0 +1,340 @@
+/*
+Copyright 2021.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	cassdcapi "github.com/k8ssandra/cass-operator/operator/pkg/apis/cassandra/v1beta1"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"strconv"
+	"strings"
+	"time"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	api "github.com/k8ssandra/k8ssandra-operator/api/v1alpha1"
+)
+
+// StargateReconciler reconciles a Stargate object
+type StargateReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+// FIXME check rbac rules
+//+kubebuilder:rbac:groups=k8ssandra.io,resources=stargates,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=k8ssandra.io,resources=stargates/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=k8ssandra.io,resources=stargates/finalizers,verbs=update
+//+kubebuilder:rbac:groups=cassandra.datastax.com,namespace="k8ssandra",resources=cassandradatacenters,verbs=get;list;watch
+
+func (r *StargateReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+
+	// Fetch the Stargate instance
+	stargate := &api.Stargate{}
+	err := r.Get(ctx, req.NamespacedName, stargate)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			logger.Info("Stargate resource not found. Ignoring since object must be deleted")
+			return ctrl.Result{}, nil
+		}
+		logger.Error(err, "Failed to get Stargate")
+		return ctrl.Result{}, err
+	}
+
```

Comment:
Before doing anything else you should do `stargate = stargate.DeepCopy()`. Reads, i.e., `Get()` hit a shared cached. We don't want to modify objects in the cache.

Reply:
Ok I will. That said, the controller doesn't change the `Stargate` object, it only updates the `Deployment` object. But that might change in the future I guess, so better safe than sorry.